File: 0216_Combination Sum III/CombinationSumIII.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        DFS(1,k,n,new ArrayList<Integer>(),ans);
        return ans;
    }
    
    private void DFS(int s, int k, int n, List<Integer> curr, List<List<Integer>> ans)
    {
        if(n == 0){
            if(curr.size() == k) ans.add(new ArrayList<Integer>(curr));
            return;
        }
        
        for(int i=s;i<=9;i++){
            if(i > n) break;
            curr.add(i);
            DFS(i+1,k,n-i,curr,ans);
            curr.remove(curr.size()-1);
        }
    }
}

File: 0216_Combination Sum III/CombinationSumIII.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ans;
        
        // 2^9, generate all combinations of [1 .. 9]
        for (int i = 0; i < (1 << 9); ++i) {
            vector<int> cur;
            int sum = 0;
            // Use j if (j - 1)-th bit is 1
            for (int j = 1; j <= 9; ++j)
                if (i & (1 << (j - 1))) {
                    sum += j;
                    cur.push_back(j);
                }
            if (sum == n && cur.size() == k) 
                ans.push_back(cur);
        }
        
        return ans;
    }
};